### PR TITLE
fix(datatrak): RN-1321: Prevent bug where a user cannot access datatrak if their refresh token has become invalid

### PR DESCRIPTION
**Pull request by rohan-bes**
### Issue RN-1321:

The underlying issue with this bug is that when we first create a refresh token for a login we upsert by: user, platform, and device name. The device name can clash however (for browsers it's just the browser versions, so using datatrak on multiple browsers can cause this issue).

There were a few different ways to approach fixing this:
1. Add a unique identifier into the device name so that we don't get clashes (downside of this is that we rapidly start accruing old refresh tokens that will never get re-used)
2. Gracefully handle the refresh token clash by clearing the session from the server and the client's browser
3. Refactoring refresh token auth to be done in the orchestration-server rather than in the central-server (this is a bit of a complex change)

I went with option 2 since it felt the best balance of pros and cons, but would be interested if the reviewer thought of a different approach or preferred either 1 or 3.

### CODE REVIEW

File: packages/datatrak-web-server/src/routes/UserRoute.ts [hunk 1]
```diff
@@ -50,7 +50,7 @@ export class UserRoute extends Route<UserRequest> {
       project = projects.find((p: WebServerProjectRequest.ResBody) => p.id === projectId);
     }
     if (countryId) {
-      const countryResponse = await ctx.services.central.fetchResources(`/entities/${countryId}`, {
+      const countryResponse = await ctx.services.central.fetchResources(`entities/${countryId}`, {
```
Comment (rohan-bes) [inline packages/datatrak-web-server/src/routes/UserRoute.ts:53]: This is unrelated, and it doesn't change behaviour. It just removes a double `/` from the url which was bothering me.

File: packages/datatrak-web/src/AppProviders.tsx [hunk 2]
```diff
@@ -12,23 +12,29 @@ import { theme } from './theme';
 import { Toast } from './components';
 import { errorToast } from './utils';
 import { CurrentUserContextProvider } from './api';
+import { REDIRECT_ERROR_PARAM } from './constants';
+
+const handleError = (error: any, query: any) => {
+  if (error.responseData.redirectClient) {
+    // Redirect the browser to the specified URL and display the error
+    window.location.href = `${error.responseData.redirectClient}?${REDIRECT_ERROR_PARAM}=${error.message}`;
```
Comment (rohan-bes) [inline packages/datatrak-web/src/AppProviders.tsx:20]: This refreshes the tab which is why we need to pass through the error as a url param. I spent a good while trying to get it to work by navigating to `/login` with `react-router-dom` but it was proving quite tricky. Ultimately I think I prefer this solution anyway since it really gives the sense of the user being booted out and forced to login again.

Comment (alexd-bes) [reply packages/datatrak-web/src/AppProviders.tsx:20]: Out of interest, what went wrong here with using `react-router-dom`?

Comment (rohan-bes) [reply packages/datatrak-web/src/AppProviders.tsx:20]: I think in pre-v6 you could directly manipulate the browser history using `window.history.pushState({}, '', url)`. But with v6 and the use of `BrowserRouter` it uses its own history, so you can't directly change it outside of a component.

Since we want the error handling to come from the `QueryClient` which is defined outside of a component, it was a bit awkward trying to inject in a `useNavigate()` reference.

I was able to do that in the end, but even then it didn't seem to be navigating to `/login`. There's probably a bit more work that could be done to work out exactly this is the case, but I think I was finding other issues due to state not being cleared. Forcing a tab refresh felt a lot more reliable as a way of kicking the user out.

Comment (rohan-bes) [reply packages/datatrak-web/src/AppProviders.tsx:20]: I think maybe because the `CurrentUserContext` still had my user? So we kept getting sent back to the home screen by the `AuthViewLoggedInRedirect`?

File: packages/datatrak-web/src/api/CurrentUserContext.tsx [hunk 3]
```diff
@@ -27,10 +26,6 @@ export const CurrentUserContextProvider = ({ children }: { children: React.React
     return <FullPageLoader />;
   }
 
-  if (currentUserQuery.isError) {
-    return <ErrorDisplay title="Error loading user" error={currentUserQuery.error as Error} />;
```
Comment (alexd-bes) [inline packages/datatrak-web/src/api/CurrentUserContext.tsx:31]: Just checking, are there ever going to be cases where this is still needed? E.g. some other kind of error?

**Review by alexd-bes — APPROVED** [review]
LGTM -  I had a small question in there but otherwise seems like a good solution for the error handling

File: packages/datatrak-web/src/api/CurrentUserContext.tsx (hunk 3 above)
Comment (rohan-bes) [reply packages/datatrak-web/src/api/CurrentUserContext.tsx:31]: Tbh I'm not sure, for instance the request could timeout or something. But this is true of any request and I feel the standard behaviour of just showing a toast error is cleaner than what we do here.

Comment (alexd-bes) [reply packages/datatrak-web/src/api/CurrentUserContext.tsx:31]: I think what I mean here is that what this `ErrorDisplay` does is hides all content in case there are some things that are visible. I think it also solved an issue with re-rendering/infinite loops. Potentially could we instead redirect to the login page for all user errors?
